chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,84 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T<..>).
+    /// Implemented(!T<..>: A).
+    /// Implemented(!T<..>: B).
+    /// Implemented(!T<..>: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T<..>` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        let interner = builder.interner();
+        let substitution = builder.substitution_in_scope();
+        let alias = AliasTy::Opaque(OpaqueTy {
+            opaque_ty_id: self.opaque_ty_id,
+            substitution: substitution.clone(),
+        });
+
+        let alias_ty = Ty::new(
+            interner,
+            ApplicationTy {
+                name: TypeName::OpaqueType(self.opaque_ty_id),
+                substitution,
+            },
+        );
+
+        builder.push_binders(&self.bound, |builder, opaque_ty_bound| {
+            // AliasEq(T<..> = HiddenTy) :- Reveal.
+            builder.push_clause(
+                DomainGoal::Holds(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: opaque_ty_bound.hidden_ty.clone(),
+                    }
+                    .cast(interner),
+                ),
+                iter::once(DomainGoal::Reveal(())),
+            );
+
+            // AliasEq(T<..> = !T<..>).
+            builder.push_fact(DomainGoal::Holds(
+                AliasEq {
+                    alias: alias.clone(),
+                    ty: alias_ty.clone(),
+                }
+                .cast(interner),
+            ));
+
+            for bound in &opaque_ty_bound.bounds {
+                // Implemented(!T<..>: Bound).
```

Comment:
PS, thanks for the comments! Even minimal stuff like this really helps to make it easier to track what's going on.